### ISSUE TITLE
Client with open cursor query can't be used for any other queries

### ISSUE THREAD

**Issue by migueloller** [brianc/node-postgres#1895]
When using a client to read from a table using a cursor, all subsequent queries to that client hang forever unless the cursor is closed.

```js
const { Pool } = require('pg')
const Cursor = require('pg-cursor')

async function main({ env: { DATABASE_URL } }) {
  const pool = new Pool({ connectionString: DATABASE_URL })

  const client = await pool.connect()

  try {
    const cursor = client.query(new Cursor('SELECT * FROM generate_series(1, 10)'))

    await client.query('SELECT 1') // This hangs forever because the cursor is still open...
  } finally {
    client.release()
  }

  await pool.end()
}

main(process).catch(error => {
  console.error(error)
  process.exit(1)
})
```

**Comment by migueloller** [comment]
> @migueloller I've run into the same issue. What did you end up doing? Using two clients?

Yup—this is specially insidious when attempting to use transactions. You might want to share a client so you can do things transactionally but that would result in a timeout since once you get a cursor, you can't do another query until the cursor is released.

And it makes sense, a cursor is stateful so the client is busy until it's released, similar to why queries for a transaction have to go through the same client.

So yeah, you'll want to make sure you always have a separate client connection when opening a cursor.

**Comment by twelve17** [comment]
> > @migueloller I've run into the same issue. What did you end up doing? Using two clients?
> 
> Yup—this is specially insidious when attempting to use transactions. You might want to share a client so you can do things transactionally but that would result in a timeout since once you get a cursor, you can't do another query until the cursor is released.
> 
> And it makes sense, a cursor is stateful so the client is busy until it's released, similar to why queries for a transaction have to go through the same client.
> 
> So yeah, you'll want to make sure you always have a separate client connection when opening a cursor.

Yep, glad to hear what I'm doing is not too far off from what you concluded.  Thanks for the response. :)

**Comment by brianc** [comment]
Hey! Just wanted to chime in here & let ya know this is actually a behavior of postgres itself. The backend wont issue a `readyForQuery` event until the cursor has been read to the end, closed by the user, or an error happens during reading the cursor.  So that means if a client is in use reading from a cursor it cannot issue any other queries until the cursor has been read to the end or manually closed.  If you absolutely need to do an ETL like process where you read some rows, process them, and insert them all in a transaction I think your best bet is to do a loop with a series of `SELECT * FROM FOO OFFSET ${offset} LIMIT ${limit}` followed by the insert of that batch etc until you read the whole table or whatever the query returns, all inside a `begin` and `commit` section.  Not ideal since a cursor can be significantly more performant than a series of limit+offset queries but I don't know of another way.  If you find one out let me know though - always interested in tips. 😄

**Comment by migueloller** [comment]
That makes sense @brianc, thanks for the response. I imagined this could be an underlying behavior of Postgres...
